sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConsumeValueTaskCorrectly.cs

```diff
@@ -0,0 +1,201 @@
+                {
+                    var walker = new ConsumeValueTaskWalker(c.SemanticModel);
+                    walker.Visit(c.Node);
+
+                    foreach (var syntaxNodes in walker.SymbolUsages.Values)
+                    {
+                        if (syntaxNodes.Count() > 1)
+                        {
+                            c.ReportDiagnosticWhenActive(Diagnostic.Create(messageOnlyOnce, syntaxNodes.First().GetLocation(),
+                                additionalLocations: syntaxNodes.Skip(1).Select(node => node.GetLocation()).ToArray()));
+                        }
+                    }
+
+                    foreach (var node in walker.ConsumedButNotCompleted)
+                    {
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(messageOnlyIfCompletedSuccessfully, node.GetLocation()));
+                    }
+                },
+                // should also check in properties?
+                SyntaxKind.MethodDeclaration);
+        }
+
+        private class ConsumeValueTaskWalker : CSharpSyntaxWalker
+        {
+            private readonly SemanticModel semanticModel;
+
+            // The key is the 'ValueTask' variable symbol, the value is a list of nodes where it is consumed
+            public Dictionary<ISymbol, IList<SyntaxNode>> SymbolUsages { get; }
+
+            // A list of 'ValueTask' nodes on which '.Result' or '.GetAwaiter().GetResult()' has been invoked when the operation has not yet completed
+            public List<SyntaxNode> ConsumedButNotCompleted { get; }
+
+            public ConsumeValueTaskWalker(SemanticModel semanticModel)
+            {
+                this.semanticModel = semanticModel;
+                SymbolUsages = new Dictionary<ISymbol, IList<SyntaxNode>>();
+                ConsumedButNotCompleted = new List<SyntaxNode>();
+            }
+
+            /**
+             * Check if 'await' is done on a 'ValueTask'
+             */
+            public override void VisitAwaitExpression(AwaitExpressionSyntax node)
+            {
+                if (node.Expression is IdentifierNameSyntax identifierName &&
+                    this.semanticModel.GetSymbolInfo(identifierName).Symbol is ISymbol symbol &&
+                    symbol.GetSymbolType().OriginalDefinition.IsAny(ValueTaskTypes))
+                {
+                    AddToSymbolUsages(symbol, identifierName);
+                }
+
+                base.VisitAwaitExpression(node);
+            }
+
+            /**
+             * Check if it's the wanted method on a ValueTask
+             * - we treat AsTask() like await - always add it to the list
+             * - for GetAwaiter().GetResult() - ignore the call if it's called inside an 'if (valueTask.IsCompletedSuccessfully)'
+             */
+            public override void VisitInvocationExpression(InvocationExpressionSyntax node)
+            {
+                if (node.Expression is MemberAccessExpressionSyntax memberAccess)
```

Comment:
it will be easier to read if you use helper methods to match method signature `AsTask` and `GetResult` like `SonarAnalyzer.Helpers.CSharpSyntaxHelper.IsMethodInvocation` (and add another helper method for the `Result` property)